plugins/mac-virtualcam/src/obs-plugin/plugin-main.mm

```diff
@@ -9,33 +12,164 @@
 	return "macOS virtual webcam output";
 }
 
+NSString *const OBSDalDestination = @"/Library/CoreMediaIO/Plug-Ins/DAL";
+
+static bool cmio_extension_supported()
+{
+	if (@available(macOS 13.0, *)) {
+		return true;
+	} else {
+		return false;
+	}
+}
+
 struct virtualcam_data {
 	obs_output_t *output;
 	obs_video_info videoInfo;
 	CVPixelBufferPoolRef pool;
+
+	// CMIO Extension (available since macOS 12.3
+	CMSimpleQueueRef queue;
+	CMIODeviceID deviceID;
+	CMIOStreamID streamID;
+	CMFormatDescriptionRef formatDescription;
+	id extensionDelegate;
+
+	// Legacy DAL (deprecated since macOS 12.3)
 	OBSDALMachServer *machServer;
 };
 
-static bool check_dal_plugin()
+@interface SystemExtensionActivationDelegate
+	: NSObject <OSSystemExtensionRequestDelegate> {
+@private
+	struct virtualcam_data *_vcam;
+}
+
+@property (getter=isInstalled) BOOL installed;
+@property NSString *lastErrorMessage;
+- (instancetype)init __unavailable;
+@end
+
+@implementation SystemExtensionActivationDelegate
+
+- (id)initWithVcam:(virtualcam_data *)vcam
+{
+	self = [super init];
+
+	if (self) {
+		_vcam = vcam;
+		_installed = NO;
+	}
+
+	return self;
+}
+
+- (OSSystemExtensionReplacementAction)
+			    request:(nonnull OSSystemExtensionRequest *)request
+	actionForReplacingExtension:
+		(nonnull OSSystemExtensionProperties *)existing
+		      withExtension:(nonnull OSSystemExtensionProperties *)ext
+{
+	NSString *extVersion =
+		[NSString stringWithFormat:@"%@.%@", [ext bundleShortVersion],
+					   [ext bundleVersion]];
+	NSString *existingVersion = [NSString
+		stringWithFormat:@"%@.%@", [existing bundleShortVersion],
+				 [existing bundleVersion]];
+
+	if ([extVersion compare:existingVersion
+			options:NSNumericSearch] == NSOrderedDescending) {
+		return OSSystemExtensionReplacementActionReplace;
+	} else {
+		return OSSystemExtensionReplacementActionCancel;
+	}
+}
+
+- (void)request:(nonnull OSSystemExtensionRequest *)request
+	didFailWithError:(nonnull NSError *)error
+{
+	NSString *errorMessage;
+	int severity;
+
+	switch (error.code) {
+	case OSSystemExtensionErrorRequestCanceled:
+		errorMessage =
+			@"macOS Camera Extension installation request cancelled.";
+		severity = (int)LOG_INFO;
+		break;
+	case OSSystemExtensionErrorUnsupportedParentBundleLocation:
+		self.lastErrorMessage = [NSString
+			stringWithUTF8String:
+				obs_module_text(
+					"Error.SystemExtension.WrongLocation")];
+		errorMessage = self.lastErrorMessage;
+		severity = (int)LOG_ERROR;
+		break;
+	default:
+		self.lastErrorMessage = error.localizedDescription;
+		errorMessage = [NSString
+			stringWithFormat:
+				@"OSSystemExtensionErrorCode %ld (\"%s\")",
+				error.code,
+				error.localizedDescription.UTF8String];
+		severity = (int)LOG_ERROR;
```

Comment:
Nit: `LOG_ERROR` is meant for "places that should definitely not fail". Failure in these places is kinda expected, so I'd go for `LOG_WARNING` instead.

Reply:
`OSSystemExtensionErrorRequestCanceled` is not an error per-se because it is based on user interaction.

`OSSystemExtensionErrorUnsupportedParentBundleLocation` is an error, because you cannot use system extensions outside of the Applications directory.

Anything else is an unhandled situation and as such should always be treated as an error, because "you should usually not arrive at this point" and if you do it warrants the severity.

Reply:
> `OSSystemExtensionErrorUnsupportedParentBundleLocation` is an error, because you cannot use system extensions outside of the Applications directory.

What I'm referring to is the definition of purely the definition of `LOG_ERROR`, not necessarily an error in the sense of "something with the extension went wrong".

> Use if there's a problem that can potentially affect the program,
> but isn't enough to require termination of the program.
> 
> Use in creation functions and core subsystem functions.  Places that
> should definitely not fail.

I would argue that the fact that we're gracefully handling this, going as far as giving a localized response to it, makes it an expected-to-happen state (especially since every user will run into it at least once, until they allowed the extension in system settings).

> Anything else is an unhandled situation and as such should always be treated as an error, because "you should usually not arrive at this point" and if you do it warrants the severity.

Yeah you're right about these, anything else (wrong code signature etc) definitely shouldn't happen.

Reply:
You convinced me about the default branch being something that definitely shouldn't happen, I'd agree with your initial assessment about it being error-worthy.